[YouTube] Support richGridRenderer on channel page to fix extraction of videos in channels

Theta-Dev opened the pull request:
YouTube is currently A/B testing a new layout on their channel pages, which uses a RichGridRenderer. The current version of NewPipe cant extract any videos from it, resulting in an empty channel page. This PR adds support for it.

Here is a response file with the new layout: https://code.thetadev.de/ThetaDev/rustypipe/src/branch/main/testfiles/channel/channel_videos_20221011_richgrid2.json

I could not reproduce the error on my own without tools (the A/B test seems to be pretty rare for me). But there are already people having issues with the channel pages in NewPipe. See https://libreddit.kavin.rocks/r/NewPipe/comments/y1zg5m/why_are_certain_educative_channels_invisible_from/

- [X] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.
- [X] I have tested the API against [NewPipe](https://github.com/TeamNewPipe/NewPipe).
- [ ] I agree to create a pull request for [NewPipe](https://github.com/TeamNewPipe/NewPipe) as soon as possible to make it compatible with the changed API (no API changes).

Fixes TeamNewPipe/NewPipe#9223

Theta-Dev commented:
Here is a debug build for testing:

[Uploading app-debug.zip…]()

Theta-Dev commented:
Note that this new channel layout includes up to 3 tabs for videos, livestreams and shorts. So we need a way to fetch up to 3 tabs at once to fully support this.

Lets hope YouTube is not going to make this permanent...

![Screenshot from 2022-10-18 23-38-32](https://user-images.githubusercontent.com/19636565/196552732-7b95f036-6021-4dd9-b06b-fc184855f81e.png)

AudricV commented:
> Lets hope YouTube is not going to make this permanent...

It will be, as it has been announced by YouTube. Check out [this video of the Creator Insights official YouTube channel](https://www.youtube.com/watch?v=czIyqEC4V-s).

> So we need a way to fetch up to 3 tabs at once to fully support this.

I don't think that's a good idea, we should *just* implement channel tabs in the extractor instead.

Theta-Dev commented:
I guess you are right. I was thinking about having to use up to 3 requests go get subscription feed data, but that wont work anyway. YouTube does not show the upload date of shorts, so it wont be possible to merge the 3 tabs into a chronological feed.

So I guess we should stick to using the RSS feed for fetching the subscription feed (maybe make that the default option once channel tabs are permanently established).

For the actual channel views we would need the 3tab layout in NewPipe (and by that occasion, add channel playlists, too).

AudricV commented on extractor/src/test/java/org/schabi/newpipe/extractor/services/youtube/YoutubeChannelExtractorTest.java:
```diff
@@ -648,4 +649,100 @@ public void testVerified() throws Exception {
+            defaultTestRelatedItems(extractor);
+        }
+
+        @Override
+        public void testMoreRelatedItems() throws Exception {
+        }
```
Why is this test empty?

Theta-Dev replied:
When I wrote this, I did not know how to reproduce YouTube's A/B tests. So I edited the mock response manually and did not add another one for continuation items. I will fix that.

AudricV reviewed (commented):
Thanks, I just have a question.

TobiGr commented on extractor/src/test/java/org/schabi/newpipe/extractor/services/youtube/YoutubeChannelExtractorTest.java:
```diff
@@ -648,4 +649,100 @@ public void testVerified() throws Exception {
+        @BeforeAll
+        public static void setUp() throws Exception {
+            YoutubeTestsUtils.ensureStateless();
+            NewPipe.init(DownloaderFactory.getDownloaderOfType(RESOURCE_PATH + "richGrid", DownloaderType.MOCK));
```
Why is the mock downloader needed here?

Theta-Dev replied:
Because the new layout is A/B tested, I had to use the mock downloader to reproduce it.

In my channel tabs PR I added the option to manually set the visitor data cookie which allows you to reproduce A/B tests without using the mock downloader.

AudricV commented:
@Theta-Dev Could you answer our questions, please? We'd like to merge this fix ASAP.

Theta-Dev commented:
Okay, I added the `setVisitorData` method from my channel tabs PR to reproduce the A/B test that way.

Also redownloaded the mock files and switched the test to a channel (Jeff Geerling) that has all 3 tabs.

AudricV commented on extractor/src/test/java/org/schabi/newpipe/extractor/services/youtube/YoutubeChannelExtractorTest.java:
```diff
@@ -648,4 +649,102 @@ public void testVerified() throws Exception {
             assertFalse(extractor.isVerified());
         }
     }
+
+    /**
+     * Test the extraction of channel data from a RichGridRenderer (currently A/B tested, as of 11.10.2022)
+     */
+    public static class RichGrid implements BaseChannelExtractorTest {
```
I am not sure if a test with a new channel is needed, as the VSauce channel has the three tabs:

<img src="https://user-images.githubusercontent.com/74829229/198875516-b4cd9a99-dc96-47b8-8920-d07951fd8eb1.png" width="600px" alt="VSauce_channel_new_channel_tabs">

AudricV reviewed (changes requested):
Almost good!

acrodemocide commented on extractor/src/test/java/org/schabi/newpipe/extractor/services/youtube/YoutubeChannelExtractorTest.java:
```diff
@@ -648,4 +649,102 @@ public void testVerified() throws Exception {
+
+        @Test
+        public void testDescription() throws Exception {
+            ExtractorAsserts.assertContains("K8s, Ansible, Raspberry Pi, photography, tech, DIY, tools", extractor.getDescription());
```
Are we worried about whether the description changes at all? I figure they usually don't, but I wasn't sure how these tests need to be maintained when the description on this channel is changed. (Unless we're working with the owner of this channel and he's agreed not to change the description).

gokulaknnan commented:
> 

Debug.apk is working on mi mobile

SebiderSushi commented:
> YouTube does not show the upload date of shorts, so it wont be possible to merge the 3 tabs into a chronological feed.

I'm accessing YouTube from germany. The shorts tab does not display an upload date in UI. However, if i <code>curl&nbsp;https://www.youtube.com/c/JeffGeerling/shorts</code> the JSON page data does contain one.
<details>
  <summary><b>Fuzzy timestamp in a richItemRenderer JSON tree</b></summary>
("vor 10 Monaten" means "10 months ago" in german)
<pre>
"richItemRenderer":{
  "content":{
    "reelItemRenderer":{
      "navigationEndpoint":{
        "reelWatchEndpoint":{
          "overlay":{
            "reelPlayerOverlayRenderer":{
              "reelPlayerHeaderSupportedRenderers":{
                "reelPlayerHeaderRenderer":{
                  "timestampText":{
                    "simpleText":"vor 10 Monaten"
                  },
                  "accessibility":{
                    "accessibilityData":{
                      "label":"What is the Log4J vulnerability? #Shorts Jeff Geerling vor 10 Monaten"
                    }
                  }
</pre>
</details>

Also, are ["Channel Uploads Playlists"](https://www.youtube.com/playlist?list=UUR-DXc1voovS8nhAvccRZhg) useful here in any capacity?
When youtube chooses to present me with the new three-tab channel layout these still contain all video types in chronological order and the JSON page data contains a fuzzy upload date for every item.
<details>
  <summary><b>Fuzzy timestamp in a playlistVideoRenderer JSON tree</b></summary>
("vor 5 Tagen" is "5 days ago" in german)
<pre>
"playlistVideoRenderer":{
  "title":{
    "accessibility":{
      "accessibilityData":{
        "label":"Better than Netflix: Jellyfin on my NAS von Jeff Geerling vor 5 Tagen 14 Minuten, 44 Sekunden 232.885 Aufrufe"
      }
    }
  },
  "videoInfo":{
    "runs":[
      {"text":"232.885 Aufrufe"},
      {"text":" • "},
      {"text":"vor 5 Tagen"}
    ]
  }
</pre>
</details>

AudricV commented:
As the change seems to be globally rolled out, I removed the `visitorData` changes and the addition of your channel test (see why here: https://github.com/TeamNewPipe/NewPipeExtractor/pull/944#discussion_r1008834846). I also updated all the channel mocks, which contains these `richGridRenderer`s.

This doesn't mean that I didn't like your code, please don't take this personally (especially because this is the third time that I do this in your PRs): I find the `visitorData` changes useful, but as they are not needed to fix this issue, I think it is better to open a separate PR for them!

Stypox reviewed (approved):
Thank you! I could not find anything to say about the code

plusterkopp commented:
sorry if this is a dumb question, but when can we expect a new NewPipe build based on this fix?

yauhen-l commented:
Hi,
thanks everyone for the effort.
Seems like fix does not work for me:

## Exception
* __User Action:__ searched
* __Request:__ фейгин лайф
* __Content Country:__ US
* __Content Language:__ en-US
* __App Language:__ en_US
* __Service:__ YouTube
* __Version:__ 0.24.1
* __OS:__ Linux Android 13 - 33
<details><summary><b>Crash log </b></summary><p>

```
org.schabi.newpipe.extractor.exceptions.ParsingException: Could not get subscriber count
	at org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeChannelInfoItemExtractor.getSubscriberCount(YoutubeChannelInfoItemExtractor.java:84)
	at org.schabi.newpipe.extractor.channel.ChannelInfoItemsCollector.extract(ChannelInfoItemsCollector.java:40)
	at org.schabi.newpipe.extractor.MultiInfoItemsCollector.extract(MultiInfoItemsCollector.java:84)
	at org.schabi.newpipe.extractor.MultiInfoItemsCollector.extract(MultiInfoItemsCollector.java:48)
	at org.schabi.newpipe.extractor.InfoItemsCollector.commit(InfoItemsCollector.java:104)
	at org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeSearchExtractor.collectStreamsFrom(YoutubeSearchExtractor.java:226)
	at org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeSearchExtractor.getInitialPage(YoutubeSearchExtractor.java:162)
	at org.schabi.newpipe.extractor.utils.ExtractorHelper.getItemsPageOrLogError(ExtractorHelper.java:21)
	at org.schabi.newpipe.extractor.search.SearchInfo.getInfo(SearchInfo.java:69)
	at org.schabi.newpipe.extractor.search.SearchInfo.getInfo(SearchInfo.java:37)
	at org.schabi.newpipe.util.ExtractorHelper.lambda$searchFor$0(ExtractorHelper.java:82)
	at org.schabi.newpipe.util.ExtractorHelper.$r8$lambda$qyxpuXgomWa-cbONQns-pd7zxm0(Unknown Source:0)
	at org.schabi.newpipe.util.ExtractorHelper$$ExternalSyntheticLambda6.call(Unknown Source:8)
	at io.reactivex.rxjava3.internal.operators.single.SingleFromCallable.subscribeActual(SingleFromCallable.java:43)
	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4813)
	at io.reactivex.rxjava3.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89)
	at io.reactivex.rxjava3.core.Scheduler$DisposeTask.run(Scheduler.java:644)
	at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:65)
	at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:56)
	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:307)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
	at java.lang.Thread.run(Thread.java:1012)
Caused by: org.schabi.newpipe.extractor.utils.Parser$RegexException: Failed to find pattern "([\d]+([\.,][\d]+)?)" inside of "@FeyginLive"
	at org.schabi.newpipe.extractor.utils.Parser.matchGroup(Parser.java:78)
	at org.schabi.newpipe.extractor.utils.Parser.matchGroup(Parser.java:63)
	at org.schabi.newpipe.extractor.utils.Parser.matchGroup1(Parser.java:52)
	at org.schabi.newpipe.extractor.utils.Utils.mixedNumberWordToLong(Utils.java:79)
	at org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeChannelInfoItemExtractor.getSubscriberCount(YoutubeChannelInfoItemExtractor.java:81)
	... 23 more

```
</details>
<hr>
 
 This is how channel looks like in the App:
![image](https://user-images.githubusercontent.com/2591672/213404296-53b76c45-8651-4e9b-a840-3c3a5ccb0d85.png)
